### Assessing Data
Inspecting data set for two things: data quality issues and lack of tidiness
- Quality Isssues means content issues like missing, duplicate, or incorrect data (so called dirty data)
- Untidy Data has specific structural issues (so called messy data)

### Types of Assessing data
- Visual Assessment
- Programmatic Assessment

### Case Study -- Auralin Phase II Clinical Trial Dataset
This Auralin Phase II clinical trial dataset comes in three tables: patients, treatments, and adverse_reactions. Acquaint yourself with them through visual assessment below.

### Gather

In [2]:
import pandas as pd
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

### Visual Assessment
In the cells below, each column of each table in this clinical trial dataset is described. To see the table that goes hand in hand with these descriptions, display each table in its entirety by displaying the pandas DataFrame that it was gathered into. This task is the mechanical part of visual assessment in pandas.

In [3]:
# Display the patients table
patients.columns

Index(['patient_id', 'assigned_sex', 'given_name', 'surname', 'address',
       'city', 'state', 'zip_code', 'country', 'contact', 'birthdate',
       'weight', 'height', 'bmi'],
      dtype='object')

patients columns:

- patient_id: the unique identifier for each patient in the Master Patient Index (i.e. patient database) of the pharmaceutical company that is producing Auralin
- assigned_sex: the assigned sex of each patient at birth (male or female)
- given_name: the given name (i.e. first name) of each patient
- surname: the surname (i.e. last name) of each patient
- address: the main address for each patient
- city: the corresponding city for the main address of each patient
- state: the corresponding state for the main address of each patient
- zip_code: the corresponding zip code for the main address of each patient
- country: the corresponding country for the main address of each patient (all United states for this clinical trial)
- contact: phone number and email information for each patient
- birthdate: the date of birth of each patient (month/day/year). The inclusion criteria for this clinical trial is age >= 18 (there is no maximum age because diabetes is a growing problem among the elderly population)
- weight: the weight of each patient in pounds (lbs)
- height: the height of each patient in inches (in)
- bmi: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m2 where kg is a person's weight in kilograms and m2 is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. The inclusion criteria for this clinical trial is 16 >= BMI >= 38.

In [4]:
# Display the treatments table
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

350 patients participated in this clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before. All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After four weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:
- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks


treatments columns:

- given_name: the given name of each patient in the Master Patient Index that took part in the clinical trial
- surname: the surname of each patient in the Master Patient Index that took part in the clinical trial
- auralin: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) and the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the international unit of measurement and the standard measurement for insulin.
- novodra: same as above, except for patients that continued treatment with Novodra
- hba1c_start: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The HbA1c test measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.
- hba1c_end: the patient's HbA1c level at the end of the last week of treatment
- hba1c_change: the change in the patient's HbA1c level from the start of treatment to the end, i.e., hba1c_start - hba1c_end. For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

In [7]:
# Display the adverse_reactions table
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

adverse_reactions columns:

- given_name: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- surname: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- adverse_reaction: the adverse reaction reported by the patient

Additional useful information:

- Insulin resistance varies person to person, which is why both starting median daily dose and ending median daily dose are required, i.e., to calculate change in dose.
- It is important to test drugs and medical products in the people they are meant to help. People of different age, race, sex, and ethnic group must be included in clinical trials. This diversity is reflected in the patients table.
- Ensuring column names are descriptive enough is an important step in acquainting yourself with the data. 'Descriptive enough' is subjective. Ideally you want short column names (so they are easier to type and read in code form) but also fully descriptive. Length vs. descriptiveness is a tradeoff and common debate (a similar debate exists for variable names). The auralin and novodra column names are probably not descriptive enough, but you'll address that later so don't worry about that for now.

### Data Quality Dimensions

Data quality dimensions help guide your thought process while assessing and also cleaning. The four main data quality dimensions are:

- ##### Completeness:
do we have all of the records that we should? Do we have missing records or not? Are there specific rows, columns, or cells missing? (e.g.Missing HbA1c changes)

- ##### Validity: 
we have the records, but they're not valid, i.e., they don't conform to a defined schema. A schema is a defined set of rules for data. These rules can be real-world constraints (e.g. negative height is impossible) and table-specific constraints (e.g. unique key constraints in tables). (e.g. Zipcode has four digitals sometimes)

- ##### Accuracy: 
inaccurate data is wrong data that is valid. It adheres to the defined schema, but it is still incorrect. Example: a patient's weight that is 5 lbs too heavy because the scale was faulty. (e.g. 27 in height)

- ##### Consistency: 
inconsistent data is both valid and accurate, but there are multiple correct ways of referring to the same thing. Consistency, i.e., a standard format, in columns that represent the same data across tables and/or within tables is desired. (e.g. NY and New York)

### Programmatic Assessment

In [13]:
patients.sample(5)

patient_id assigned_sex given_name     surname                   address  \
448         449         male       Ivan       Fomin         632 Peaceful Lane   
473         474       female       Kate   Wilkinson           664 Lyon Avenue   
211         212       female     Martha  Afanasyeva      3613 Lodgeville Road   
172         173         male      Alvin     Jackson  3411 Pyramid Valley Road   
15           16         male      Søren        Lund    2438 Shady Pines Drive   

                 city state  zip_code        country  \
448  Garfield Heights    OH   44128.0  United States   
473      South Boston    MA    2127.0  United States   
211        Saint Paul    MN   55114.0  United States   
172         Iowa City    IA   52240.0  United States   
15          Kingsport    VA   37660.0  United States   

                                       contact  birthdate  weight  height  \
448           216-502-3773IvanFomin@dayrep.com  6/10/1930   139.9      65   
473    KateWilkinson@armyspy.com1 508 905 2371  7/18/1998   175.3      65   
211  MarthaAfanasyeva@dayrep.com1 612 228 4170  10/3/1997   151.2      73   
172      319-541-3295AlvinAJackson@armyspy.com  12/1/1945   177.8      71   
15             276-225-1955SrenFLund@gustr.com  8/23/1922   201.5      64   

      bmi  
448  23.3  
473  29.2  
211  19.9  
172  24.8  
15   34.6

In [12]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [14]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [15]:
adverse_reactions['adverse_reaction'].value_counts()

hypoglycemia                 19
injection site discomfort     6
headache                      3
nausea                        2
cough                         2
throat irritation             2
Name: adverse_reaction, dtype: int64

In [18]:
patients[patients['address'].isnull()]
# some missing contact info

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [19]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

- 75% and max hb1c change is above 90%, which is massive while the 50% is only 34%, that suggests a massive skew.
- When we check the 0.97 is for Elliot Richardson and 7.56 - 7.09 is 0.47. it's simply calculated wrong.

### Data Quality Summary

#### Patients table
- Zipcode is a float not a string
- Zipcode has four digitals sometimes
- Tim Neudorf (no.4) height is 27 instaed of 72 is posssible
- Full state names sometimes, abbreviations other times (e.g. NY and New York)
- Missing demographic information (address -contact info)
- Erroneous data types (assigned sex, state, zip_code, birthdate)
- Multiple phone number formats

#### Treatments table:
- Missing HbA1c changes
- The letter 'u' in starting and ending does for Auralin and Novodra (need to transfer to inter for further calculate)
- Lowercase given names and surnames (problem in joining tables)
- Missing records (280 instead of 350)
- Erroneous datatype (auralin and novodra columns)
- Inaccurate HbA1c change (Elliot Richardson)

#### Adverse_reactions table:
- lowercase given names and surnames